torch/testing/_internal/common_device_type.py

```diff
@@ -474,30 +495,30 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
     # Acquires members names
     # See Note [Overriding methods in generic tests]
     generic_members = set(generic_test_class.__dict__.keys()) - set(empty_class.__dict__.keys())
-    generic_tests = [x for x in generic_members if x.startswith('test')]
+    generic_tests = [x for x in generic_members if x.startswith('test')]\
+
+    # Filter out the device types based on user inputs
+    desired_device_types = filter_desired_device_types(DEFAULT_DEVICE_TYPE_LIST,
+                                                       except_for, only_for)
 
-    def split_if_not_empty(x):
+    def split_if_not_empty(x: str):
         return x.split(",") if len(x) != 0 else []
 
-    # Derive defaults from environment variables if available, default is still none
+    # Filter out the device types based on environment variables if available
     # Usage:
     # export PYTORCH_TESTING_DEVICE_ONLY_FOR=cuda,cpu
     # export PYTORCH_TESTING_DEVICE_EXCEPT_FOR=xla
-    if only_for is None:
-        only_for = split_if_not_empty(os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", ''))
+    env_only_for = split_if_not_empty(os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", ''))
+    env_except_for = split_if_not_empty(os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", ''))
+
+    desired_device_types = filter_desired_device_types(desired_device_types,
+                                                       env_except_for, env_only_for)
 
-    if except_for is None:
-        except_for = split_if_not_empty(os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", ''))
 
     # Creates device-specific test cases
     for base in device_type_test_bases:
-        # Skips bases listed in except_for
-        if except_for and only_for:
-            assert base.device_type not in except_for or base.device_type not in only_for,\
-                "same device cannot appear in except_for and only_for"
-        if except_for and base.device_type in except_for:
-            continue
-        if only_for and base.device_type not in only_for:
+        # Skips bases if not listed in desired_device_types
+        if base.device_type not in desired_device_types:
```

Comment:
I would look at applying the filter here so you don't need DEFAULT_DEVICE_TYPE_LIST.